src/Features/CSharp/Portable/Completion/KeywordRecommenders/BoolKeywordRecommender.cs

```diff
@@ -35,6 +35,8 @@ protected override bool IsValidContext(int position, CSharpSyntaxContext context
                 context.IsPrimaryFunctionExpressionContext ||
                 context.IsCrefContext ||
                 syntaxTree.IsAfterKeyword(position, SyntaxKind.ConstKeyword, cancellationToken) ||
+                syntaxTree.IsAfterKeyword(position, SyntaxKind.RefKeyword, cancellationToken) ||
```

Comment:
I'm surprised this doesn't fall under a more abstract category, like "IsTypeContext" or "IsPossibleTypeContext" (those are made-up, I'm not sure whether such methods exist). If there is not one, consider factoring one out, or filing an issue to do so.
Note: this would avoid duplicating the fix to multiple recommenders.

@dotnet/roslyn-ide for advice. Is there already a check that would be suitable for this?

Reply:
One more thought on this. Is there a reason that all the recommenders for intrinsic types (bool, byte, string, object, etc) don't just share code?

---
In reply to: [175509814](https://github.com/dotnet/roslyn/pull/25570#discussion_r175509814) [](ancestors = 175509814)

Reply:
@jcouv I was really surprised by this too and asked @CyrusNajmabadi who said this is very much by design. I would still like to do some cleanup here and reduce the duplication to some degree but I'm not going to do that as part of this PR.

Reply:
@jcouv @CyrusNajmabadi said that for starters there are some cases (enum base list, stackalloc) where only some of those keywords are allowed, that's why they can't share everything. That said, I still think the duplication could be reduced significantly, but I'm doing this PR as a bugfix rather than code quality improvement/refactoring.

Reply:
It woud require refactoring some of the existing helpers to be more fine-grained to support this sort of distinction. Yes, there is a helper called IsTypeContext but it is more general and offered in more places than these keywords, not less - so it's not as easy as using the helper and then adding more checks with `||`.

Reply:
but @CyrusNajmabadi has a lot more to say about this than I do

Reply:
> I'm surprised this doesn't fall under a more abstract category, like "IsTypeContext" or "IsPossibleTypeContext" (those are made-up, I'm not sure whether such methods exist). If there is not one, consider factoring one out, or filing an issue to do so.

We can consider these sorts of things. However, as i mentioned to @Neme12 it gets a little subtle due to different types being legal in different contexts.  

> One more thought on this. Is there a reason that all the recommenders for intrinsic types (bool, byte, string, object, etc) don't just share code?

Yes.  Not all of them are allowed in the same contexts.  For example, after stackalloc, you can have types, but not all types.

Similarly, you can have integral types (but not floating point types) in an enum-base-list.

--

Now, as i mentioned to @Neme12 we could certainly extract out some contexts that were well defined, and would make sense for large sets of these providers.  For example: IsIntegralTypeContext (for non-floating point numeric types).

Reply:
Thanks for the clarification. I certainly agree it doesn't need to be solved in this PR.

Reply:
@Neme12 has shown a great interest in improving things here based on supporting all our existing goals, while also massively reducing duplication.  I'm really excited to see what we can do here!